File: sort/merge-sort/java/MergeSort.java
import java.util.*;

public class MergeSort {
    public static int[] sort(int[] A) {
	if (A.length < 2) return A;
	int n = A.length;
	int[] left = Arrays.copyOfRange(A, 0, n/2);
	int[] right = Arrays.copyOfRange(A, n/2, n);
	int[] sorted_left = MergeSort.sort(left);
	int[] sorted_right = MergeSort.sort(right);
	return merge(sorted_left, sorted_right);
    }

    private static int[] merge(int[] A, int[] B) {
	int n = A.length;
	int m = B.length;
	int[] C = new int[n+m];
	int i = 0, j = 0, k = 0;
	while (i < n && j < m) {
	    if (A[i] < B[j]) {
		C[k++] = A[i++];
	    } else {
		C[k++] = B[j++];
	    }
	}
	while (i < n) {
	    C[k++] = A[i++];
	}
	while (j < m) {
	    C[k++] = B[j++];
	}
	return C;
    }
}

File: sort/mergesort/merge_sort.cpp
#include <algorithm>
#include <vector>
#include "merge_sort.h"

void Merge(const std::vector<int>& arr1,
           const std::vector<int>& arr2,
           std::vector<int>* out) {
  int m = arr1.size();
  int n = arr2.size();
  out->resize(m + n);
  int i = 0, j = 0, k = 0;
  while (i < m && j < n) {
    if (arr1[i] < arr2[j]) {
      (*out)[k++] = arr1[i++];
    } else {
      (*out)[k++] = arr2[j++];
    }
  }
  while (i < m) {
    (*out)[k++] = arr1[i++];
  }
  while (j < n) {
    (*out)[k++] = arr2[j++];
  }
}

void Mergesort(const std::vector<int>& arr, std::vector<int>* out) {
  out->resize(arr.size());
  if (arr.size() < 2) {
    copy(arr.begin(), arr.end(), out->begin());
    return;
  }
  int middle = arr.size() / 2;
  std::vector<int> left(middle);
  copy(arr.begin(), arr.begin()+middle, left.begin());
  std::vector<int> right(arr.size() - middle);
  copy(arr.begin()+middle, arr.end(), right.begin());
  std::vector<int> sorted_left, sorted_right;
  Mergesort(left, &sorted_left);
  Mergesort(right, &sorted_right);
  Merge(sorted_left, sorted_right, out);
}
